opm/autodiff/CPRPreconditioner.hpp

```diff
@@ -67,209 +72,138 @@ template<class M, class X, class Y, class P>
-createILUnPtr(const M& A, int ilu_n, double relax,
-              const Dune::OwnerOverlapCopyCommunication<I1,I2>& comm)
+template<class M, class X, class C>
+std::shared_ptr<ParallelOverlappingILU0<M,X,X,C> >
+createILUnPtr(const M& A, const C& comm, int ilu_n, double relax)
 {
-    typedef Dune::BlockPreconditioner<
-        X,
-        X,
-        Dune::OwnerOverlapCopyCommunication<I1,I2>,
-        Dune::SeqILUn<M,X,X>
-        > PointerType;
-    Dune::SeqILUn<M,X,X>* ilu = new Dune::SeqILUn<M,X,X>( A, ilu_n, relax);
-
-    return typename SelectParallelILUSharedPtr<Dune::SeqILUn<M,X,X>, I1, I2>::type
-        (new PointerType(*ilu, comm),createParallelDeleter(*ilu, comm));
+    return std::make_shared<ParallelOverlappingILU0<M,X,X,C> >( A, comm, ilu_n, relax);
 }
-#endif
-
-/// \brief Creates the elliptic preconditioner (ILU0)
-/// \param Ae The matrix of the elliptic system.
-/// \param relax The relaxation parameter for ILU0
-template<class M, class X=typename M::range_type>
-std::unique_ptr<Dune::SeqILU0<M,X,X> >
-createEllipticPreconditionerPointer(const M& Ae, double relax,
-                                    const Dune::Amg::SequentialInformation&)
-{
-    return std::unique_ptr<Dune::SeqILU0<M,X,X> >(new Dune::SeqILU0<M,X,X>(Ae, relax));
-}
-
-#if HAVE_MPI
 /// \brief Creates the elliptic preconditioner (ILU0)
 /// \param Ae    The matrix of the elliptic system.
 /// \param relax The relaxation parameter for ILU0.
 /// \param comm  The object describing the parallelization information and communication.
-template<class M, class X=typename M::range_type, class I1, class I2>
-typename CPRSelector<M,X,X,Dune::OwnerOverlapCopyCommunication<I1,I2> >
-::EllipticPreconditionerPointer
+template<class M, class X=typename M::range_type, class P>
+typename CPRSelector<M,X,X,P>::EllipticPreconditionerPointer
 createEllipticPreconditionerPointer(const M& Ae, double relax,
-                                    const Dune::OwnerOverlapCopyCommunication<I1,I2>& comm)
+                                    const P& comm)
 {
-    typedef typename CPRSelector<M,X,X,Dune::OwnerOverlapCopyCommunication<I1,I2> >
+    typedef typename CPRSelector<M,X,X,P >
         ::EllipticPreconditioner ParallelPreconditioner;
 
-    //Dune::SeqILU0<M,X,X>* ilu=new Dune::SeqILU0<M,X,X>(Ae, relax);
-    typedef typename CPRSelector<M,X,X,Dune::OwnerOverlapCopyCommunication<I1,I2> >
+    typedef typename CPRSelector<M,X,X,P>
         ::EllipticPreconditionerPointer EllipticPreconditionerPointer;
     return EllipticPreconditionerPointer(new ParallelPreconditioner(Ae, comm, relax));
 }
 
-#endif // #if HAVE_MPI
+template < class C, class Op, class P, class S, std::size_t index >
+inline void
+createAMGPreconditionerPointer(Op& opA, const double relax, const P& comm,
+                               std::unique_ptr< BlackoilAmg<Op,S,C,P,index> >& amgPtr,
+                               const CPRParameter& params = CPRParameter())
```

Comment:
I get lots of compile errors trying to compile this PR. The first one is here: invalid use of incomplete type 'Opm::CPRParameter'. That error is repeated multiple times. I also get: no matching function for call to 'createAMGPreconditionerPointer', and similar errors. Forgotten #include?

Reply:
Weired that my compilers (g++ 6.3.18 and 4.9.2) did not complain. But yours is right constructing CPRParameter needs the definition which only available at the end of the file.

Reply:
Weired that my compilers (g++ 6.3.18 and 4.9.2) did not complain. But yours is right constructing CPRParameter needs the definition which only available at the end of the file.

Reply:
Fixed.

Reply:
At least that is what I guess. Please try again.